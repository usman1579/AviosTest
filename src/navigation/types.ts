import type {CompositeScreenProps, NavigatorScreenParams} from '@react-navigation/native';
import type {BottomTabScreenProps} from '@react-navigation/bottom-tabs';
import type {StackScreenProps} from '@react-navigation/stack';
// import type {DrawerNavigationProp} from '@react-navigation/drawer';

export const MAIN_ROUTES = {
  MAIN: 'Main',
  HOME: 'Home',
  SPEND: 'Spend',
  COLLECT: 'Collect',
  ACCOUNT: 'Account',
  SPLASH: 'Splash',
  APP_LOADING: 'AppLoading',
  SIGN_IN: 'SignIn',
  SIGN_UP: 'SignUp',
} as const;

export type DrawerParamList = {
  [MAIN_ROUTES.HOME]: undefined;
};

export type TabStackParamList = {
  [MAIN_ROUTES.HOME]: undefined;
  [MAIN_ROUTES.SPEND]: undefined;
  [MAIN_ROUTES.COLLECT]: undefined;
  [MAIN_ROUTES.ACCOUNT]: undefined;
};

export type StackParamList = {
  [MAIN_ROUTES.MAIN]: NavigatorScreenParams<TabStackParamList> | undefined;
  [MAIN_ROUTES.SPLASH]: undefined;
  [MAIN_ROUTES.APP_LOADING]: undefined;
  [MAIN_ROUTES.SIGN_IN]: undefined;
  [MAIN_ROUTES.SIGN_UP]: undefined;
};

export type StackScreenProp<T extends keyof StackParamList> = StackScreenProps<StackParamList, T>;

export type TabScreenProp<T extends keyof TabStackParamList> = CompositeScreenProps<
  BottomTabScreenProps<TabStackParamList, T>,
  StackScreenProp<keyof StackParamList>
>;

// export type GlobalScreenProp = CompositeScreenProps<>

// export type GlobalNavigationProp = CompositeNavigationProp<
//   BottomTabNavigationProp<TabStackParamList>,
//   CompositeNavigationProp<
//     StackNavigationProp<StackParamList>,
//     DrawerNavigationProp<DrawerParamList>
//   >
// >;
