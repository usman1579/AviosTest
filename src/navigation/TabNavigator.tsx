import React from 'react';
import {createBottomTabNavigator} from '@react-navigation/bottom-tabs';
import Ionicons from 'react-native-vector-icons/Ionicons';

import {HomeScreen} from '@/screens/HomeScreen';
import {SpendScreen} from '@/screens/SpendScreen';
import {CollectScreen} from '@/screens/CollectScreen';
import {AccountScreen} from '@/screens/AccountScreen';
import {TabStackParamList, MAIN_ROUTES} from './types';

const Tab = createBottomTabNavigator<TabStackParamList>();

function SetTabBarIcon(routeName: string, focused: boolean, color: string, size: number) {
  let iconName = '';
  if (routeName === 'Home') {
    iconName = focused ? 'home-sharp' : 'home-outline';
  } else if (routeName === 'Spend') {
    iconName = focused ? 'airplane' : 'airplane-outline';
  } else if (routeName === 'Collect') {
    iconName = focused ? 'wallet' : 'wallet-outline';
  } else if (routeName === 'Account') {
    iconName = focused ? 'person' : 'person-outline';
  }

  return <Ionicons name={iconName} size={size} color={color} />;
}

export function TabNavigator() {
  return (
    <Tab.Navigator
      screenOptions={({route}) => ({
        tabBarIcon: ({focused, color, size}) => SetTabBarIcon(route.name, focused, color, size),
        tabBarActiveTintColor: '#075AAA',
        tabBarInactiveTintColor: 'black',
      })}>
      <Tab.Screen name={MAIN_ROUTES.HOME} component={HomeScreen} />
      <Tab.Screen name={MAIN_ROUTES.SPEND} component={SpendScreen} />
      <Tab.Screen name={MAIN_ROUTES.COLLECT} component={CollectScreen} />
      <Tab.Screen name={MAIN_ROUTES.ACCOUNT} component={AccountScreen} />
    </Tab.Navigator>
  );
}
