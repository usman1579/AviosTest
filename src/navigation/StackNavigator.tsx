import React from 'react';
import {createStackNavigator} from '@react-navigation/stack';
import {TabNavigator} from './TabNavigator';
import {useAppSelector} from '@/store/hooks';
import {selectIsAuthenticated} from '@/slices/auth';
import {StackParamList, MAIN_ROUTES} from './types';
import {SplashScreen} from '@/screens/SplashScreen';
import {AppLoadingScreen} from '@/screens/AppLoadingScreen';
import {SignInScreen} from '@/screens/SignInScreen';
import {SignUpScreen} from '@/screens/SignUpScreen';

const Stack = createStackNavigator<StackParamList>();

export function StackNavigator() {
  const isLoggedIn = useAppSelector(selectIsAuthenticated);

  return (
    <Stack.Navigator screenOptions={{headerShown: false}}>
      {isLoggedIn && (
        <>
          <Stack.Screen name={MAIN_ROUTES.APP_LOADING} component={AppLoadingScreen} />
          <Stack.Screen name="Main" component={TabNavigator} />
        </>
      )}
      {!isLoggedIn && (
        <>
          <Stack.Screen name={MAIN_ROUTES.SPLASH} component={SplashScreen} />
          <Stack.Screen name={MAIN_ROUTES.SIGN_IN} component={SignInScreen} />
          <Stack.Screen name={MAIN_ROUTES.SIGN_UP} component={SignUpScreen} />
        </>
      )}
    </Stack.Navigator>
  );
}
