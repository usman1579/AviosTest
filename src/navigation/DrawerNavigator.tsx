import React from 'react';
import {createDrawerNavigator} from '@react-navigation/drawer';
import {StackNavigator} from './StackNavigator';
import {DrawerParamList, MAIN_ROUTES} from './types';

const Drawer = createDrawerNavigator<DrawerParamList>();

export function DrawerNavigator() {
  return (
    <Drawer.Navigator
      useLegacyImplementation={false}
      screenOptions={{
        // swipeEnabled: false,
        drawerStyle: {
          width: '100%',
        },
        drawerPosition: 'right',
        headerShown: false,
        drawerType: 'back',
        unmountOnBlur: true,
      }}>
      <Drawer.Screen name={MAIN_ROUTES.HOME} component={StackNavigator} />
    </Drawer.Navigator>
  );
}
