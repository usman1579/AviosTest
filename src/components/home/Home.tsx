import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { useAppSelector } from '@/store/hooks';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { selectMember } from '@/slices/auth';
import { MAIN_ROUTES } from '@/navigation/types';

const WELCOMEBACK = 'Welcome back';
const BALANCE = 'Balance';

export function Home({ navigation }: any) {
  const member = useAppSelector(selectMember);
  const onPress = () => navigation.navigate(MAIN_ROUTES.ACCOUNT);

  return (
    <View style={styles.root}>
      <Text style={styles.welcome}>{WELCOMEBACK}</Text>
      <Text style={styles.firstName}>
        {member?.firstName + '  '}
        <MaterialCommunityIcons name="hand-wave-outline" size={24} />
      </Text>

      <TouchableOpacity testID={'TouchPress'} onPress={onPress} style={styles.barcomponent}>
        <View style={styles.view1}>
          <Ionicons name="cube-outline" size={24} color={'black'} />
        </View>
        <View style={styles.view2}>
          <Text style={styles.text}>
            {BALANCE} <Text style={styles.balance}>{member?.avios}</Text>
          </Text>
        </View>
        <View style={styles.view3}>
          <Ionicons name="chevron-forward" size={24} color={'blue'} />
        </View>
      </TouchableOpacity>
    </View>
  );
}

export const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: 'white',
    padding: 20,
  },
  welcome: {
    fontSize: 16,
    color: 'black',
    fontWeight: '500',
  },
  firstName: {
    fontSize: 26,
    fontWeight: 'bold',
    marginVertical: 10,
  },
  barcomponent: {
    height: 70,
    width: '100%',
    backgroundColor: 'white',
    alignSelf: 'center',
    borderRadius: 10,
    marginVertical: 10,
    flexDirection: 'row',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  view1: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  view2: {
    flex: 4,
    justifyContent: 'center',
    marginHorizontal: 10,
  },
  view3: {
    flex: 1,
    justifyContent: 'center',
  },
  text: {
    fontSize: 16,
    fontWeight: '500',
  },
  balance: {
    fontSize: 16,
    fontWeight: 'bold',
  },
});
