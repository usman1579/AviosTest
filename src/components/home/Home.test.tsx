import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Home } from './Home'; // Adjust the import path based on your project structure
import { MAIN_ROUTES } from '@/navigation/types';

// Mocking the navigation prop
const mockNavigation = {
  navigate: jest.fn(),
};

// Mock the useSelector hook
jest.mock('@/store/hooks', () => ({
  useAppSelector: jest.fn(),
}));

describe('Home component', () => {
  // Reset the mock before each test
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    // Mock the useSelector hook to return a dummy member
    jest.mock('@/slices/auth', () => ({
      selectMember: jest.fn(),
    }));

    // Import the mocked Home component after mocking the useSelector hook
    const { toJSON } = render(<Home navigation={mockNavigation} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('navigates to ACCOUNT screen on press', () => {
    // Mock the useSelector hook to return a dummy member
    jest.mock('@/slices/auth', () => ({
      selectMember: jest.fn(() => ({ firstName: 'John', avios: 100 })), //dummy data as needed
    }));

    const { getByTestId } = render(<Home navigation={mockNavigation} />);
    const touchableButton = getByTestId('TouchPress');

    fireEvent.press(touchableButton);

    expect(mockNavigation.navigate).toHaveBeenCalledWith(MAIN_ROUTES.ACCOUNT);
  });
});
