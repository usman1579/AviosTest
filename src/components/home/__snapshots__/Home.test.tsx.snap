// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 16,
        "fontWeight": "500",
      }
    }
  >
    Welcome back
  </Text>
  <Text
    style={
      {
        "fontSize": 26,
        "fontWeight": "bold",
        "marginVertical": 10,
      }
    }
  >
    undefined  
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      󱠢
    </Text>
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "white",
        "borderRadius": 10,
        "elevation": 5,
        "flexDirection": "row",
        "height": 70,
        "marginVertical": 10,
        "opacity": 1,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
        "width": "100%",
      }
    }
    testID="TouchPress"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "black",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        {
          "flex": 4,
          "justifyContent": "center",
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 16,
            "fontWeight": "500",
          }
        }
      >
        Balance
         
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        />
      </Text>
    </View>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "blue",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
